### PR TITLE
Http client adapters

### PR DESCRIPTION
**Pull request by jubianchi**
This is a draft implementation of HTTP adapters solving #49. 

The library now supports:
- Guzzle 3
- Guzzle 4
- Buzz

The user has to choose an adapter when creating the API client :

``` php
<?php

$buzz = new \Gitlab\HttpClient\Adapter\BuzzAdapter();
$guzzle3 = new \Gitlab\HttpClient\Adapter\Guzzle3Adapter();
$guzzle4 = new \Gitlab\HttpClient\Adapter\Guzzle4Adapter();

$bgitlab = new \Gitlab\Client('http://git.foobar.com/api/v3/', $buzz);
$bgitlab->authenticate('foobar', \Gitlab\HttpClient\HttpClientInterface::AUTH_HTTP_TOKEN);

$g3gitlab = new \Gitlab\Client('http://git.foobar.com/api/v3/', $guzzle3);
$g3gitlab->authenticate('foobar', \Gitlab\HttpClient\HttpClientInterface::AUTH_HTTP_TOKEN);

$g4gitlab = new \Gitlab\Client('http://git.foobar.com/api/v3/', $guzzle4);
$g4gitlab->authenticate('foobar', \Gitlab\HttpClient\HttpClientInterface::AUTH_HTTP_TOKEN);
```

:warning: This will break user's code when updating the library :warning: but I think it's a minor change.

### CODE REVIEW

**Comment by digitalkaoz** [comment]
why not use https://github.com/egeloen/ivory-http-adapter here? it already handles everything related and is able to guess the best matching adapter

**Comment by jubianchi** [comment]
@digitalkaoz seems to be a really good alternative! :+1:

**Comment by m4tthumphrey** [comment]
As posted in another issue:

There will not be a major new tag until the HTTP adaptor (#5, #50, #49), PHP version (#54) and PSR-4 (#55) discussions are concluded. I will also be doing a lot of house keeping over the coming days and weeks, which may include several breaking changes, mainly todo with models (#8, #10). I am not sure on the roadmap of Gitlab but envisage this to coincide with the release of 8.0 depending on @randx and co's plans.

However, I will tag a smaller release to coincide with the release of 7.8 next week.

**Comment by m4tthumphrey** [comment]
I like the look of [Ivory](https://github.com/egeloen/ivory-http-adapter) as it handles everything that we don't need to care about (ie Gitlab functionality only).

**Comment by rossedman** [comment]
@digitalkaoz @m4tthumphrey Wow! Didn't even know about Ivory. Really nice.

**Comment by Baachi** [comment]
@m4tthumphrey Any progress here? If you want i can integrate [HTTPlug](http://php-http.readthedocs.org/en/latest/httplug/introduction.html) which is a psr-7 compatible http layer which supports:
 
 * cURL
 * Socket
 * Guzzle5
 * Guzzle6
 * React

If you say yes, i can send you a PR tomorrow. (I need that for a project anyways)

**Comment by m4tthumphrey** [comment]
Crack on @Baachi. Probably best to create a new PR though!

**Comment by sagikazarmark** [comment]
@Baachi any progress with HTTPlug?

**Comment by Baachi** [comment]
@sagikazarmark Hey, i started to work on it but have some problems with the listeners.

**Comment by m4tthumphrey** [comment]
@Baachi I dont mind if some of the API has to change to support this. The client hasn't had a major (breaking) release for a long while.

**Comment by fbourigault** [comment]
@Baachi do you plan to finish this PR?

**Comment by m1guelpf** [comment]
@jubianchi I think you should close this in favour of #191

**Comment by fbourigault** [comment]
Superseeded by #191.

File: composer.json [hunk 1]
```diff
@@ -21,10 +21,12 @@
...
 	],
 	"require": {
-		"php":                ">=5.3.2",
-		"ext-curl":           "*",
-		"kriswallsmith/buzz": ">=0.7"
+		"php":                ">=5.3.2"
 	},
+    "suggest": {
+        "kriswallsmith/buzz": "To use Buzz as the HTTP adapter (~0.7)",
+        "guzzlehttp/guzzle": "To use Guzzle as the HTTP adapter (~3.7, ~4.1)"
```
[the remark this answers is not in the document]
Comment (glensc) [reply composer.json:28]: @sstok


🚧 Never link to branches 🚧
- see https://news.ycombinator.com/item?id=8046710
  and https://help.github.com/articles/getting-permanent-links-to-files/
- hit `y`
- and perhaps https://github.com/josephfrazier/octopermalinker

[the remark this answers is not in the document]
Comment (sstok) [reply composer.json:28]: Thanks for the tip, but this issue is really old 🤪 I'm no longer using Gush (or Gitlab for that matter).

**Pull request closed without merging**